src/raftstore/store/snap.rs

```diff
@@ -618,6 +665,987 @@ mod v1 {
+                }
+
+                // Check each cf file has been fully written, and the checksum matches.
+                {
+                    let mut file = cf_file.file.take().unwrap();
+                    try!(file.flush());
+                }
+                if cf_file.written_size != cf_file.size {
+                    return Err(io::Error::new(ErrorKind::Other,
+                                              format!("snapshot file {} for cf {} size \
+                                                       mismatches, real size {}, expected size \
+                                                       {}",
+                                                      cf_file.path.display(),
+                                                      cf_file.cf,
+                                                      cf_file.written_size,
+                                                      cf_file.size)));
+                }
+                let checksum = cf_file.write_digest.as_ref().unwrap().sum32();
+                if checksum != cf_file.checksum {
+                    return Err(io::Error::new(ErrorKind::Other,
+                                              format!("snapshot file {} for cf {} checksum \
+                                                       mismatches, real checksum {}, expected \
+                                                       checksum {}",
+                                                      cf_file.path.display(),
+                                                      cf_file.cf,
+                                                      checksum,
+                                                      cf_file.checksum)));
+                }
+
+                try!(fs::rename(&cf_file.tmp_path, &cf_file.path));
+                total_size += cf_file.size;
+            }
+            // write meta file
+            let mut v = vec![];
+            try!(self.meta_file.meta.write_to_vec(&mut v));
+            try!(self.meta_file.file.take().unwrap().write_all(&v[..]));
+            try!(fs::rename(&self.meta_file.tmp_path, &self.meta_file.path));
+            let mut size_track = self.size_track.wl();
+            *size_track = size_track.saturating_add(total_size);
+            Ok(())
+        }
+
+        fn apply(&mut self, options: ApplyOptions) -> Result<()> {
+            for cf_file in &mut self.cf_files {
+                if cf_file.size == 0 {
+                    // Skip empty cf file.
+                    continue;
+                }
+
+                try!(check_abort(&options.abort));
+                let cf_handle = box_try!(rocksdb::get_cf_handle(&options.db, &cf_file.cf));
+                if plain_file_used(cf_file.cf) {
+                    let mut file = box_try!(File::open(&cf_file.path));
+                    try!(apply_plain_cf_file(&mut file, &options, cf_handle));
+                } else {
+                    let ingest_opt = IngestExternalFileOptions::new().move_files(true);
```

Comment:
It's an efficient way to ingest this file.  
What failure need to be recoverable and how to make it recoverable? I'm open for any better idea.